src/coreclr/src/vm/eventpipesession.cpp

```diff
@@ -557,6 +557,9 @@ void EventPipeSession::SuspendWriteEvent()
     }
     CONTRACTL_END;
 
+    // Pause the session so that additional writes from newly created threads cannot happen after this is called.
+    Pause();
```

Comment:
I don't think we should take a dependency on the `Pause`/`Resume` APIs that were added. The backing value is explicitly not protected behind any locks and I'm not sure we should invest in  enhancing it.  It was only added for the GC gen0 promotion analysis work that went in right at the end of 5.

https://github.com/dotnet/runtime/blob/fd094a92cdcd7ccf3ae1efad5922b9a303bf44b9/src/coreclr/src/vm/eventpipe.cpp#L537-L571

Shouldn't the session already be disabled by the above logic (`EventPipe::DisableInternal`) before we reach this code path?  We remove all the providers from the session, so new event writes shouldn't be hitting this code path right?  By the time we've reached where the assertion was happening, new threads shouldn't be writing to this session.

Reply:
You're right. I checked again and it seems like the problem is happening because SetSessionWriteInProgress is set regardless of whether we actually write to the session here:

```
        for (uint32_t i = 0; i < MaxNumberOfSessions; ++i)
        {
            if ((s_allowWrite & ((uint64_t)1 << i)) == 0)
                continue;

            // Now that we know this session is probably live we pay the perf cost of the memory barriers
            // Setting this flag lets a thread trying to do a concurrent disable that it is not safe to delete
            // session ID i. The if check above also ensures that once the session is unpublished this thread
            // will eventually stop ever storing ID i into the WriteInProgress flag. This is important to
            // guarantee termination of the YIELD_WHILE loop in SuspendWriteEvents.
            pEventPipeThread->SetSessionWriteInProgress(i);
            {
                EventPipeSession *const pSession = s_pSessions[i].Load();

                // Disable is allowed to set s_pSessions[i] = NULL at any time and that may have occured in between
                // the check and the load
                if (pSession != nullptr)
                    pSession->WriteEvent(
                        pThread,
                        event,
                        payload,
                        pActivityId,
                        pRelatedActivityId,
                        pEventThread,
                        pStack);
            }
            // Do not reference pSession past this point, we are signaling Disable() that it is safe to
            // delete it
            pEventPipeThread->SetSessionWriteInProgress(UINT32_MAX);
```

Let me follow up with a different fix to address the issue.

Reply:
If I remember correctly, isn't the session disabled before we call SuspendWriteEvent? It seems like the real issue here is that we shouldn't be creating new threads for a disabled session

Reply:
So I thought about this for a while and I believe we still have a race that looks like this:

1. <Thread 1> Calls EventPipe::WriteEvent and is pre-emptied [here](https://github.com/dotnet/runtime/blob/814efd9470e326f20d32bd7b9b2ab1c2fbe8cb34/src/coreclr/vm/eventpipe.cpp#L842) before it calls `SetSessionWriteInProgress`.

2. <Thread 2> Calls EventPipeSession::SuspendWriteEvent.

3. <Thread 2> Skips through the YIELD_WHILE for Thread 1 because `SetSessionWriteInProgress` hasn't been called by Thread 1.

4. <Thread 1> Wakes up and calls `SetSessionWriteInProgress` and starts writing

5. <Thread 1> Gets preemptied before `SetSessionWriteInProgress(UINT32_MAX) is called`

6. <Thread 2> Calls EventPipeBufferManager::SuspendWriteEvent() which sees that Thread 1's `SessionWriteinProgress` is still `sessionIndex`.

Adding the YIELD_WHILE in EventPipeBufferManager::SuspendWriteEvent() should prevent this situation from happening.

Reply:
I'll make some time tonight to dig in and actually code review tonight